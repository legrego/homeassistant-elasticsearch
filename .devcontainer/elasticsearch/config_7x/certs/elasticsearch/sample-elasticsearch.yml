#
# SAMPLE ELASTICSEARCH CONFIGURATION FOR ENABLING SSL ON THE HTTP INTERFACE
#
# This is a sample configuration snippet for Elasticsearch that enables and configures SSL for the HTTP (Rest) interface
#
# This was automatically generated at: 2020-12-21 13:19:13-05:00
# This configuration was intended for Elasticsearch version 8.0.0
#
# You should review these settings, and then update the main configuration file at
#   /Users/larry/repos/kibana/.es/8.0.0/config/elasticsearch.yml
#

# This turns on SSL for the HTTP (Rest) interface
xpack.security.http.ssl.enabled: true

# This configures the keystore to use for SSL on HTTP
xpack.security.http.ssl.keystore.path: "http.p12"

