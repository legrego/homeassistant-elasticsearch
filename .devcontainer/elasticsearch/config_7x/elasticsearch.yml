#
# Path to directory where to store the data (separate multiple locations by comma):
#
path.data: /var/lib/elasticsearch
#
# Path to log files:
#
path.logs: /var/log/elasticsearch


cluster.name: home-assistant-cluster

xpack.security.http.ssl.enabled: true
xpack.security.http.ssl.keystore.path: certs/elasticsearch/http.p12
xpack.security.enabled: true