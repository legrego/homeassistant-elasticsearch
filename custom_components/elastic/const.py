""" constants """

DOMAIN = 'elastic'

CONF_PUBLISH_ENABLED = 'publish_enabled'
CONF_HEALTH_SENSOR_ENABLED = 'health_sensor_enabled'
CONF_INDEX_FORMAT = 'index_format'
CONF_PUBLISH_FREQUENCY = 'publish_frequency'
CONF_ONLY_PUBLISH_CHANGED = 'only_publish_changed'
CONF_ILM_ENABLED = 'ilm_enabled'
CONF_ILM_POLICY_NAME = 'ilm_policy_name'
CONF_ILM_MAX_SIZE = 'ilm_max_size'
CONF_ILM_DELETE_AFTER = 'ilm_delete_after'
CONF_SSL_CA_PATH = 'ssl_ca_path'
CONF_CLOUD_ID = 'cloud_id'

CONF_TAGS = 'tags'

ONE_MINUTE = 60
ONE_HOUR = 60 * 60

VERSION_SUFFIX = "-v4_1"

INDEX_TEMPLATE_NAME = "hass-index-template" + VERSION_SUFFIX
