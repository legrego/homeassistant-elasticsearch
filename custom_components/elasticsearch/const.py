""" constants """

DOMAIN = "elasticsearch"

CONF_PUBLISH_ENABLED = "publish_enabled"
CONF_HEALTH_SENSOR_ENABLED = "health_sensor_enabled"
CONF_INDEX_FORMAT = "index_format"
CONF_PUBLISH_FREQUENCY = "publish_frequency"
CONF_EXCLUDED_DOMAINS = "excluded_domains"
CONF_EXCLUDED_ENTITIES = "excluded_entities"
CONF_PUBLISH_MODE = "publish_mode"
CONF_INCLUDED_DOMAINS = "included_domains"
CONF_INCLUDED_ENTITIES = "included_entities"
CONF_ONLY_PUBLISH_CHANGED = "only_publish_changed"
CONF_ILM_ENABLED = "ilm_enabled"
CONF_ILM_POLICY_NAME = "ilm_policy_name"
CONF_ILM_MAX_SIZE = "ilm_max_size"
CONF_ILM_DELETE_AFTER = "ilm_delete_after"
CONF_SSL_CA_PATH = "ssl_ca_path"

CONF_TAGS = "tags"

ONE_MINUTE = 60
ONE_HOUR = 60 * 60

VERSION_SUFFIX = "-v4_1"

INDEX_TEMPLATE_NAME = "hass-index-template" + VERSION_SUFFIX

PUBLISH_MODE_ALL = "All"
PUBLISH_MODE_STATE_CHANGES = "State changes"
PUBLISH_MODE_ANY_CHANGES = "Any changes"
