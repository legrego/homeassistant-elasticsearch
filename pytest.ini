[pytest]
addopts = -rxf -x -v -l --cov=./ --cov-report=xml
filterwarnings =
    ignore::DeprecationWarning
testpaths = tests
python_files =
    test_*.py