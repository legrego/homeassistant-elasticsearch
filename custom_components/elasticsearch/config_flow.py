"""Config flow for Elastic."""

from typing import Any

import voluptuous as vol
from homeassistant import config_entries
from homeassistant.config_entries import ConfigEntry, ConfigFlowResult
from homeassistant.const import (
    CONF_API_KEY,
    CONF_PASSWORD,
    CONF_TIMEOUT,
    CONF_URL,
    CONF_USERNAME,
    CONF_VERIFY_SSL,
)
from homeassistant.core import callback
from homeassistant.helpers.selector import (
    BooleanSelector,
    BooleanSelectorConfig,
    NumberSelector,
    NumberSelectorConfig,
    SelectSelector,
    SelectSelectorConfig,
    TargetSelector,
    TargetSelectorConfig,
    TextSelector,
    TextSelectorConfig,
    TextSelectorType,
)

from custom_components.elasticsearch.const import (
    CONF_AUTHENTICATION_TYPE,
    CONF_CHANGE_DETECTION_TYPE,
    CONF_EXCLUDE_TARGETS,
    CONF_INCLUDE_TARGETS,
    CONF_POLLING_FREQUENCY,
    CONF_PUBLISH_FREQUENCY,
    CONF_SSL_CA_PATH,
    CONF_SSL_VERIFY_HOSTNAME,
    CONF_TAGS,
    CONF_TARGETS_TO_EXCLUDE,
    CONF_TARGETS_TO_INCLUDE,
    ONE_MINUTE,
    StateChangeType,
)
from custom_components.elasticsearch.const import DOMAIN as ELASTIC_DOMAIN
from custom_components.elasticsearch.errors import (
    AuthenticationRequired,
    CannotConnect,
    InsufficientPrivileges,
    UntrustedCertificate,
)
from custom_components.elasticsearch.es_gateway_8 import Elasticsearch8Gateway

from .logger import LOGGER as BASE_LOGGER
from .logger import (
    async_log_enter_exit_debug,
    async_log_enter_exit_info,
    log_enter_exit_debug,
    log_enter_exit_info,
)

CONFIG_TO_REDACT = {CONF_API_KEY, CONF_PASSWORD, CONF_USERNAME}

# Data Flow values
SCHEMA_URL: dict[Any, Any] = {"schema": CONF_URL, "default": "https://localhost:9200"}
SCHEMA_AUTHENTICATION_TYPE: dict[Any, Any] = {
    "schema": CONF_AUTHENTICATION_TYPE,
    "default": CONF_API_KEY,
}
SCHEMA_USERNAME = {"schema": CONF_USERNAME, "default": None}
SCHEMA_PASSWORD = {"schema": CONF_PASSWORD, "default": None}
SCHEMA_VERIFY_SSL = {"schema": CONF_VERIFY_SSL, "default": True}
SCHEMA_SSL_VERIFY_HOSTNAME = {"schema": CONF_SSL_VERIFY_HOSTNAME, "default": True}
SCHEMA_SSL_CA_PATH: dict[str, Any] = {"schema": CONF_SSL_CA_PATH}
SCHEMA_API_KEY = {"schema": CONF_API_KEY, "default": None}
SCHEMA_TIMEOUT = {"schema": CONF_TIMEOUT, "default": 30}


TRANSLATION_KEY_STATE = "state"
TRANSLATION_KEY_ATTRIBUTE = "attribute"
TRANSLATION_KEY_API_KEY = "api_key"
TRANSLATION_KEY_BASIC_AUTH = "basic_auth"


class ElasticFlowHandler(config_entries.ConfigFlow, domain=ELASTIC_DOMAIN):
    """Handle an Elastic config flow."""

    VERSION = 7
    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
    GATEWAY = Elasticsearch8Gateway

    def __init__(self) -> None:
        """Initialize the Elastic flow."""
        self._reauth_entry: ConfigEntry | None = None
        self._prospective_config: dict[str, Any] = self.init_data or {}

    @async_log_enter_exit_info
    async def async_step_user(
        self, user_input: dict | None = None, errors: dict | None = None
    ) -> ConfigFlowResult:  # noqa: ARG002
        """Handle a flow initialized by the user. This is the first step in the flow.

        We will gather the url of the elasticsearch cluster and the desired authentication method.
        """

        if errors is not None:
            BASE_LOGGER.debug("async_step_user errors: %s", errors)

        if user_input is not None:
            # If the URL has an https schema, test the connection and see if we get an untrusted certificate error
            prospective_settings: dict = {CONF_URL: user_input.get(CONF_URL)}
            self._prospective_config.update(user_input)

            try:
                await Elasticsearch8Gateway.async_init_then_stop(**prospective_settings)

            except UntrustedCertificate:
                return await self.async_step_certificate_issues()
            except CannotConnect:
                BASE_LOGGER.debug("Cannot connect", exc_info=True)
                return await self.async_step_user(errors={CONF_URL: "cannot_connect"})
            except AuthenticationRequired:
                return await self.async_step_authentication_issues()

            return self.async_step_complete()

        return self.async_show_form(
            step_id="user",
            data_schema=vol.Schema(
                {
                    vol.Required(**SCHEMA_URL): TextSelector(
                        TextSelectorConfig(
                            type=TextSelectorType.URL,
                        )
                    ),
                },
            ),
            errors=errors,
        )

    @async_log_enter_exit_info
    async def async_step_certificate_issues(
        self, user_input: dict | None = None, errors: dict | None = None
    ) -> ConfigFlowResult:
        """Check to see if we need to ask the user for more specific SSL settings."""

        if errors is not None:
            BASE_LOGGER.debug("async_step_certificate_issues errors: %s", errors)

        if user_input is not None:
            try:
                await Elasticsearch8Gateway.async_init_then_stop(
                    url=self._prospective_config[CONF_URL],
                    verify_certs=user_input.get(CONF_VERIFY_SSL, True),
                    ca_certs=user_input.get(CONF_SSL_CA_PATH),
                    verify_hostname=user_input.get(CONF_SSL_VERIFY_HOSTNAME, True),
                )

                self._prospective_config.update(user_input)

                # If we get here, we have a valid connection, so we can complete our flow
                return self.async_step_complete()

            except UntrustedCertificate:
                BASE_LOGGER.debug("Certificate issue", exc_info=True)
                return await self.async_step_certificate_issues(errors={"base": "untrusted_certificate"})

            except CannotConnect:
                BASE_LOGGER.debug("Cannot connect", exc_info=True)
                return await self.async_step_user(errors={"base": "cannot_connect"})

            except AuthenticationRequired:
                self._prospective_config.update(user_input)
                return await self.async_step_authentication_issues()

        return self.async_show_form(
            step_id="certificate_issues",
            data_schema=vol.Schema(
                {
                    vol.Required(**SCHEMA_VERIFY_SSL): BooleanSelector(
                        BooleanSelectorConfig(),
                    ),
                    vol.Required(**SCHEMA_SSL_VERIFY_HOSTNAME): BooleanSelector(
                        BooleanSelectorConfig(),
                    ),
                    vol.Optional(**SCHEMA_SSL_CA_PATH): TextSelector(
                        TextSelectorConfig(
                            type=TextSelectorType.TEXT,
                        )
                    ),
                    # To do: consider switching to a file selector to upload the ca cert vol.Optional(**SCHEMA_SSL_CA_PATH): FileSelector(FileSelectorConfig(accept=".pem,.crt")),
                },
            ),
            errors=errors,
        )

    @async_log_enter_exit_info
    async def async_step_authentication_issues(self, user_input: dict | None = None) -> ConfigFlowResult:
        """We require Authentication information to connect to Elasticsearch."""

        if user_input is not None:
            if user_input[CONF_AUTHENTICATION_TYPE] == TRANSLATION_KEY_BASIC_AUTH:
                return await self.async_step_basic_auth()
            if user_input[CONF_AUTHENTICATION_TYPE] == TRANSLATION_KEY_API_KEY:
                return await self.async_step_api_key()

        # Show a form where we ask the user if they want to use API Key or Basic Auth
        return self.async_show_form(
            step_id="authentication_issues",
            data_schema=vol.Schema(
                {
                    vol.Required(**SCHEMA_AUTHENTICATION_TYPE): SelectSelector(
                        SelectSelectorConfig(
                            translation_key="authentication_issues",
                            options=[TRANSLATION_KEY_API_KEY, TRANSLATION_KEY_BASIC_AUTH],
                        )
                    ),
                }
            ),
        )

    @async_log_enter_exit_info
    async def async_step_basic_auth(
        self, user_input: dict | None = None, errors: dict | None = None
    ) -> ConfigFlowResult:
        """Prompt the user for the settings required to use Basic Authentication against the Elasticsearch cluster."""

        if errors is not None:
            BASE_LOGGER.debug("async_step_basic_auth errors: %s", errors)

        if user_input is not None:
            try:
                await Elasticsearch8Gateway.async_init_then_stop(
                    url=self._prospective_config[CONF_URL],
                    username=user_input.get(CONF_USERNAME),
                    password=user_input.get(CONF_PASSWORD),
                    verify_certs=self._prospective_config.get(CONF_VERIFY_SSL, True),
                    verify_hostname=self._prospective_config.get(CONF_SSL_VERIFY_HOSTNAME, True),
                    ca_certs=self._prospective_config.get(CONF_SSL_CA_PATH),
                )
            except InsufficientPrivileges:
                BASE_LOGGER.debug("Insufficient Privileges", exc_info=True)
                return await self.async_step_basic_auth(
                    errors={TRANSLATION_KEY_BASIC_AUTH: "insufficient_privileges"}
                )
            except AuthenticationRequired:
                BASE_LOGGER.debug("Invalid basic authentication", exc_info=True)
                return await self.async_step_basic_auth(
                    errors={TRANSLATION_KEY_BASIC_AUTH: "invalid_basic_auth"}
                )

            # We are authenticated, update settings and complete flow
            self._prospective_config.update(user_input)
            return self.async_step_complete()

        return self.async_show_form(
            step_id="basic_auth",
            data_schema=vol.Schema(
                {
                    vol.Required(**SCHEMA_USERNAME): TextSelector(
                        TextSelectorConfig(
                            type=TextSelectorType.TEXT,
                        )
                    ),
                    vol.Required(**SCHEMA_PASSWORD): TextSelector(
                        TextSelectorConfig(
                            type=TextSelectorType.PASSWORD,
                        )
                    ),
                },
            ),
            errors=errors,
        )

    @async_log_enter_exit_info
    async def async_step_api_key(
        self, user_input: dict | None = None, errors: dict | None = None
    ) -> ConfigFlowResult:
        """Prompt the user for the settings required to use API Key Authentication against the Elasticsearch cluster."""

        if errors is not None:
            BASE_LOGGER.debug("async_step_api_key errors: %s", errors)

        if user_input is not None:
            try:
                await Elasticsearch8Gateway.async_init_then_stop(
                    url=self._prospective_config[CONF_URL],
                    api_key=user_input.get(CONF_API_KEY),
                    verify_certs=self._prospective_config.get(CONF_VERIFY_SSL, True),
                    verify_hostname=self._prospective_config.get(CONF_SSL_VERIFY_HOSTNAME, True),
                    ca_certs=self._prospective_config.get(CONF_SSL_CA_PATH),
                )
            except InsufficientPrivileges:
                BASE_LOGGER.debug("Insufficient Privileges", exc_info=True)
                return await self.async_step_api_key(
                    errors={TRANSLATION_KEY_API_KEY: "insufficient_privileges"}
                )
            except AuthenticationRequired:
                BASE_LOGGER.debug("Invalid API Key", exc_info=True)
                return await self.async_step_api_key(errors={TRANSLATION_KEY_API_KEY: "invalid_api_key"})

            # We are authenticated, update settings and complete flow
            self._prospective_config.update(user_input)
            return self.async_step_complete()

        return self.async_show_form(
            step_id="api_key",
            data_schema=vol.Schema(
                {
                    vol.Required(**SCHEMA_API_KEY): TextSelector(
                        TextSelectorConfig(
                            type=TextSelectorType.PASSWORD,
                        )
                    ),
                },
            ),
            errors=errors,
        )

    @log_enter_exit_info
    def async_step_complete(self) -> ConfigFlowResult:
        """Handle the completion of the flow."""
        default_options = ElasticOptionsFlowHandler.default_options

        if self._reauth_entry is not None:
            return self.async_update_reload_and_abort(
                self._reauth_entry,
                unique_id=self._reauth_entry.unique_id,
                title=self._reauth_entry.title,
                data=self._prospective_config,
                options={**self._reauth_entry.options},
            )

        title: str = self._prospective_config[CONF_URL]

        return self.async_create_entry(title=title, data=self._prospective_config, options=default_options)

    @staticmethod
    @callback
    def async_get_options_flow(config_entry: ConfigEntry) -> config_entries.OptionsFlow:
        """Get the options flow for this handler."""
        return ElasticOptionsFlowHandler(config_entry)

    async def async_step_reauth(self, user_input: dict | None = None) -> ConfigFlowResult:
        """Handle reauthorization."""
        entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
        if entry is None:
            return self.async_abort(reason="no_entry")

        self._reauth_entry = entry

        self._prospective_config = dict(entry.data)

        if self._prospective_config.get(CONF_USERNAME, None) is not None:
            return await self.async_step_basic_auth(user_input=user_input)
        if self._prospective_config.get(CONF_API_KEY, None) is not None:
            return await self.async_step_api_key(user_input=user_input)

        return self.async_abort(reason="no_auth")


class ElasticOptionsFlowHandler(config_entries.OptionsFlow):
    """Handle Elastic options."""

    default_options: dict[str, Any] = {
        CONF_PUBLISH_FREQUENCY: ONE_MINUTE,
        CONF_POLLING_FREQUENCY: ONE_MINUTE,
        CONF_CHANGE_DETECTION_TYPE: [StateChangeType.STATE.value, StateChangeType.ATTRIBUTE.value],
        CONF_INCLUDE_TARGETS: False,
        CONF_EXCLUDE_TARGETS: False,
        CONF_TARGETS_TO_INCLUDE: {},
        CONF_TARGETS_TO_EXCLUDE: {},
        CONF_TAGS: [],
    }

    def __init__(self, config_entry: ConfigEntry) -> None:
        """Initialize Elastic options flow."""
        self.config_entry = config_entry
        self.options = dict(config_entry.options)

    async def async_step_init(self, user_input: dict | None = None) -> ConfigFlowResult:
        """Manage the Elastic options."""

        return await self.async_step_options(user_input=user_input)

    @async_log_enter_exit_debug
    async def async_step_options(
        self,
        user_input: dict | None = None,
    ) -> ConfigFlowResult:
        """Publish Options."""
        if user_input is not None:
            self.options.update(user_input)

            self.hass.config_entries.async_schedule_reload(self.config_entry.entry_id)

            return self.async_create_entry(title="", data=self.options)

        return self.async_show_form(
            step_id="options",
            data_schema=self._build_options_schema(),
        )

    @log_enter_exit_debug
    def _build_options_schema(self) -> vol.Schema:
        """Build the options schema."""

        from_options = self.config_entry.options.get

        SCHEMA_PUBLISH_FREQUENCY = {
            "schema": CONF_PUBLISH_FREQUENCY,
            "default": from_options(CONF_PUBLISH_FREQUENCY),
        }
        SCHEMA_POLLING_FREQUENCY = {
            "schema": CONF_POLLING_FREQUENCY,
            "default": from_options(CONF_POLLING_FREQUENCY),
        }

        SCHEMA_CHANGE_DETECTION_TYPE = {
            "schema": CONF_CHANGE_DETECTION_TYPE,
            "default": from_options(CONF_CHANGE_DETECTION_TYPE),
        }
        SCHEMA_TAGS = {
            "schema": CONF_TAGS,
            "default": from_options(CONF_TAGS),
        }
        SCHEMA_INCLUDE_TARGETS = {
            "schema": CONF_INCLUDE_TARGETS,
            "default": from_options(CONF_INCLUDE_TARGETS),
        }
        SCHEMA_TARGETS_TO_INCLUDE = {
            "schema": CONF_TARGETS_TO_INCLUDE,
            "default": from_options(CONF_TARGETS_TO_INCLUDE),
        }

        SCHEMA_EXCLUDE_TARGETS = {
            "schema": CONF_EXCLUDE_TARGETS,
            "default": from_options(CONF_EXCLUDE_TARGETS),
        }
        SCHEMA_TARGETS_TO_EXCLUDE = {
            "schema": CONF_TARGETS_TO_EXCLUDE,
            "default": from_options(CONF_TARGETS_TO_EXCLUDE),
        }

        return vol.Schema(
            {
                vol.Optional(**SCHEMA_PUBLISH_FREQUENCY): NumberSelector(
                    NumberSelectorConfig(
                        min=0,
                        max=600,
                        step=10,
                        unit_of_measurement="seconds",
                    )
                ),
                vol.Optional(**SCHEMA_POLLING_FREQUENCY): NumberSelector(
                    NumberSelectorConfig(
                        min=0,
                        max=3600,
                        step=10,
                        unit_of_measurement="seconds",
                    )
                ),
                vol.Optional(**SCHEMA_CHANGE_DETECTION_TYPE): SelectSelector(
                    SelectSelectorConfig(
                        translation_key="change_detection_type",
                        options=[TRANSLATION_KEY_STATE, TRANSLATION_KEY_ATTRIBUTE],
                        multiple=True,
                    )
                ),
                vol.Optional(**SCHEMA_TAGS): SelectSelector(
                    SelectSelectorConfig(options=[], custom_value=True, multiple=True)
                ),
                vol.Optional(**SCHEMA_INCLUDE_TARGETS): BooleanSelector(
                    BooleanSelectorConfig(),
                ),
                vol.Optional(**SCHEMA_TARGETS_TO_INCLUDE): TargetSelector(
                    TargetSelectorConfig(),
                ),
                vol.Optional(**SCHEMA_EXCLUDE_TARGETS): BooleanSelector(
                    BooleanSelectorConfig(),
                ),
                vol.Optional(**SCHEMA_TARGETS_TO_EXCLUDE): TargetSelector(
                    TargetSelectorConfig(),
                ),
            }
        )
