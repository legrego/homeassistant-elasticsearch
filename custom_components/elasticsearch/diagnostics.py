"""Diagnostics for the Elasticsearch integration."""

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import (
    CONF_API_KEY,
    CONF_PASSWORD,
    CONF_USERNAME,
)
from homeassistant.core import HomeAssistant

CONFIG_TO_REDACT = {CONF_API_KEY, CONF_PASSWORD, CONF_USERNAME}


async def async_get_config_entry_diagnostics(hass: HomeAssistant, entry: ConfigEntry) -> dict[str, Any]:  # noqa: ARG001
    """Return diagnostics for the config entry."""

    return {
        "data": async_redact_data(entry.data, CONFIG_TO_REDACT),
        "options": async_redact_data(entry.options, CONFIG_TO_REDACT),
    }
