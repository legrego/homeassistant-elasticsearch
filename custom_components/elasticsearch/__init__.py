"""Support for sending event data to an Elasticsearch cluster."""

from __future__ import annotations

from logging import Logger
from typing import TYPE_CHECKING

from homeassistant.config_entries import ConfigEntry
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady, IntegrationError

from custom_components.elasticsearch.config_flow import ElasticFlowHandler
from custom_components.elasticsearch.errors import (
    AuthenticationRequired,
    CannotConnect,
    ESIntegrationException,
    UnsupportedVersion,
)
from custom_components.elasticsearch.logger import (
    LOGGER,
    async_log_enter_exit_debug,
    async_log_enter_exit_info,
    have_child,
    log_enter_exit_debug,
)

from .es_integration import ElasticIntegration

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

type ElasticIntegrationConfigEntry = ConfigEntry[ElasticIntegration]


@async_log_enter_exit_info
async def async_setup_entry(hass: HomeAssistant, config_entry: ElasticIntegrationConfigEntry) -> bool:
    """Set up integration via config flow."""

    # Create an specific logger for this config entry
    _logger: Logger = have_child(name=config_entry.title)

    _logger.info("Initializing integration for %s", config_entry.title)

    try:
        integration = ElasticIntegration(hass=hass, config_entry=config_entry, log=_logger)
        await integration.async_init()
    except (UnsupportedVersion, CannotConnect) as err:
        raise ConfigEntryNotReady(err) from err
    except AuthenticationRequired as err:
        raise ConfigEntryAuthFailed(err) from err
    except ESIntegrationException as err:
        raise ConfigEntryNotReady(err) from err
    except Exception as err:
        msg = "Unknown error occurred"
        _logger.exception(msg)
        raise IntegrationError(err) from err

    config_entry.runtime_data = integration
    return True


@async_log_enter_exit_info
async def async_unload_entry(hass: HomeAssistant, config_entry: ElasticIntegrationConfigEntry) -> bool:
    """Teardown integration."""

    if (
        hasattr(config_entry, "runtime_data")
        and config_entry.runtime_data is not None
        and isinstance(config_entry.runtime_data, ElasticIntegration)
    ):
        integration = config_entry.runtime_data

        await integration.async_shutdown()
    else:
        LOGGER.warning(
            "Called to unload config entry %s, but it doesn't appear to be loaded", config_entry.title
        )

    return True


@async_log_enter_exit_debug
async def async_migrate_entry(hass: HomeAssistant, config_entry: ElasticIntegrationConfigEntry) -> bool:
    """Handle migration of config entry."""
    if config_entry.version == ElasticFlowHandler.VERSION:
        return True

    try:
        migrated_data, migrated_options, migrated_version = migrate_data_and_options_to_version(
            config_entry,
            ElasticFlowHandler.VERSION,
        )
    except ValueError:
        LOGGER.exception(
            "Migration failed attempting to migrate from version %s to version %s. Ended on %s.",
            config_entry.version,
            ElasticFlowHandler.VERSION,
        )
        return False

    hass.config_entries.async_update_entry(
        config_entry,
        data=migrated_data,
        options=migrated_options,
        version=migrated_version,
    )

    return True


@log_enter_exit_debug
def migrate_data_and_options_to_version(
    config_entry: ElasticIntegrationConfigEntry,
    desired_version: int,
) -> tuple[dict, dict, int]:
    """Migrate a config entry from its current version to a desired version."""
    LOGGER.debug(
        "Migrating config entry from version %s to %s",
        config_entry.version,
        desired_version,
    )

    data = {**config_entry.data}
    options = {**config_entry.options}
    begin_version = config_entry.version
    current_version = begin_version

    if current_version < desired_version:
        for version in range(current_version + 1, desired_version + 1):
            migration_func = globals().get(f"migrate_to_version_{version}")
            if migration_func:
                data, options = migration_func(data, options)
                current_version = version

    end_version = current_version

    LOGGER.info("Migration from version %s to version %s successful", begin_version, end_version)

    return data, options, end_version


def migrate_to_version_2(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 2."""
    only_publish_changed = data.get("only_publish_changed", False)
    data["publish_mode"] = "All" if not only_publish_changed else "Any changes"

    if "only_publish_changed" in data:
        del data["only_publish_changed"]

    return data, options


def migrate_to_version_3(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 3."""
    if "health_sensor_enabled" in data:
        del data["health_sensor_enabled"]

    return data, options


def migrate_to_version_4(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 4."""
    if "index_mode" not in data:
        data["index_mode"] = "index"

    conf_ilm_max_size = "ilm_max_size"
    if conf_ilm_max_size in data:
        del data[conf_ilm_max_size]

    conf_ilm_delete_after = "ilm_delete_after"
    if conf_ilm_delete_after in data:
        del data[conf_ilm_delete_after]

    return data, options


def migrate_to_version_5(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 5."""
    keys_to_remove = [
        "datastream_type",
        "datastream_name_prefix",
        "datastream_namespace",
    ]

    for key in keys_to_remove:
        if key in data:
            del data[key]

    keys_to_migrate = [
        "publish_enabled",
        "publish_frequency",
        "publish_mode",
        "excluded_domains",
        "excluded_entities",
        "included_domains",
        "included_entities",
    ]

    for key in keys_to_migrate:
        if key not in options and key in data:
            options[key] = data[key]
        if key in data:
            del data[key]

    remove_keys_if_empty = [
        "username",
        "password",
        "api_key",
    ]

    for key in remove_keys_if_empty:
        if key in data and data[key] == "":
            del data[key]

    return data, options


def migrate_to_version_6(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 6."""

    if data.get("index_mode") is not None:
        del data["index_mode"]

    # Change publish mode to change_detection_type
    if options.get("publish_mode") is not None:
        if options["publish_mode"] == "All":
            options["polling_frequency"] = options["publish_frequency"]
            options["change_detection_type"] = ["STATE", "ATTRIBUTE"]

        if options["publish_mode"] == "Any changes":
            options["polling_frequency"] = 0
            options["change_detection_type"] = ["STATE", "ATTRIBUTE"]

        if options["publish_mode"] == "State changes":
            options["polling_frequency"] = 0
            options["change_detection_type"] = ["STATE"]

        del options["publish_mode"]

    else:
        options["polling_frequency"] = 0
        options["change_detection_type"] = ["STATE", "ATTRIBUTE"]

    # add dedicated settings for polling
    options_to_remove = [
        "ilm_enabled",
        "ilm_policy_name",
        "publish_mode",
        "publish_enabled",
        "index_format",
        "index_mode",
        "alias",
    ]

    for key in options_to_remove:
        if key in options:
            del options[key]

    return data, options


def migrate_to_version_7(data: dict, options: dict) -> tuple[dict, dict]:
    """Migrate config to version 7."""

    # if tags does not exist, set it to an empty array
    if "tags" not in options:
        options["tags"] = []

    options["targets_to_include"] = {}
    options["targets_to_exclude"] = {}
    options["exclude_targets"] = False
    options["include_targets"] = False

    if "included_entities" in options and len(options["included_entities"]) != 0:
        options["include_targets"] = True
        options["targets_to_include"]["entity_id"] = options["included_entities"]

    if "excluded_entities" in options and len(options["excluded_entities"]) != 0:
        options["exclude_targets"] = True
        options["targets_to_exclude"]["entity_id"] = options["excluded_entities"]

    keys_to_remove = [
        "excluded_domains",
        "excluded_entities",
        "included_domains",
        "included_entities",
    ]

    # Lowercase values in array options["change_detection_type"]
    if "change_detection_type" in options:
        options["change_detection_type"] = [x.lower() for x in options["change_detection_type"]]

    for key in keys_to_remove:
        if key in options:
            del options[key]

    return data, options
