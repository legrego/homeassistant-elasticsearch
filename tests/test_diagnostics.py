"""Tests for the Elasticsearch integration diagnostics."""

import pytest
from custom_components.elasticsearch.diagnostics import async_get_config_entry_diagnostics
from homeassistant.const import (
    CONF_API_KEY,
    CONF_PASSWORD,
    CONF_URL,
    CONF_USERNAME,
)


@pytest.mark.parametrize(
    "data",
    [
        {
            CONF_URL: "https://example.com",
            CONF_USERNAME: "test_user",
            CONF_PASSWORD: "test_password",
            CONF_API_KEY: "test_api_key",
        },
        {
            CONF_URL: "https://example.com",
            CONF_USERNAME: "test_user",
        },
        {
            CONF_URL: "https://example.com",
        },
    ],
    ids=["URL and all auth params", "Only URL and username", "Only URL"],
)
@pytest.mark.parametrize("options", [{}])
async def test_async_get_config_entry_diagnostics(hass, config_entry, data, options, snapshot):
    """Test async_get_config_entry_diagnostics function."""

    result = await async_get_config_entry_diagnostics(hass, config_entry)

    assert result == snapshot
