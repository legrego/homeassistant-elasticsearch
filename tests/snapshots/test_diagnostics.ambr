# serializer version: 1
# name: test_async_get_config_entry_diagnostics[options0-Only URL and username]
  dict({
    'data': dict({
      'url': 'https://example.com',
      'username': '**REDACTED**',
    }),
    'options': dict({
    }),
  })
# ---
# name: test_async_get_config_entry_diagnostics[options0-Only URL]
  dict({
    'data': dict({
      'url': 'https://example.com',
    }),
    'options': dict({
    }),
  })
# ---
# name: test_async_get_config_entry_diagnostics[options0-URL and all auth params]
  dict({
    'data': dict({
      'api_key': '**REDACTED**',
      'password': '**REDACTED**',
      'url': 'https://example.com',
      'username': '**REDACTED**',
    }),
    'options': dict({
    }),
  })
# ---
